src/snowflake/snowpark/modin/plugin/_internal/binary_op_utils.py

```diff
@@ -312,314 +300,527 @@ def _op_is_between_timedelta_and_numeric(
-            and isinstance(first_datatype(), _IntegralType)
-        )
-    ):
-        # string/integer case (only for mul/rmul).
-        # swap first_operand with second_operand because
-        # REPEAT(<input>, <n>) expects <input> to be string
-        if isinstance(first_datatype(), _IntegralType):
-            first_operand, second_operand = second_operand, first_operand
-
-        binary_op_result_column = iff(
-            second_operand > pandas_lit(0),
-            repeat(first_operand, second_operand),
-            # Snowflake's repeat doesn't support negative number,
-            # but pandas will return an empty string
-            pandas_lit(""),
+        if _op_is_between_two_timedeltas_or_timedelta_and_null(
+            self.first_datatype(), self.second_datatype()
+        ):
+            self.result_snowpark_pandas_type = TimedeltaType()
+        elif isinstance(
+            self.first_datatype(), TimedeltaType
+        ) and _is_numeric_non_timedelta_type(self.second_datatype()):
+            self.result_column = ceil(self.result_column)
+            self.result_snowpark_pandas_type = TimedeltaType()
+
+
+class MulOp(BinaryOp):
+    def _custom_compute(self) -> None:
+        if _op_is_between_timedelta_and_numeric(
+            self.first_datatype, self.second_datatype
+        ):
+            self.result_column = cast(
+                floor(self.first_operand * self.second_operand), LongType()
+            )
+            self.result_snowpark_pandas_type = TimedeltaType()
+        elif (
+            isinstance(self.second_datatype(), _IntegralType)
+            and isinstance(self.first_datatype(), StringType)
+        ) or (
+            isinstance(self.second_datatype(), StringType)
+            and isinstance(self.first_datatype(), _IntegralType)
+        ):
+            # string/integer case (only for mul/rmul).
+            # swap first_operand with second_operand because
+            # REPEAT(<input>, <n>) expects <input> to be string
+            if isinstance(self.first_datatype(), _IntegralType):
+                self.first_operand, self.second_operand = (
+                    self.second_operand,
+                    self.first_operand,
+                )
+
+            self.result_column = iff(
+                self.second_operand > pandas_lit(0),
+                repeat(self.first_operand, self.second_operand),
+                # Snowflake's repeat doesn't support negative number,
+                # but pandas will return an empty string
+                pandas_lit(""),
+            )
+
+
+class EqOp(BinaryOp):
+    def _custom_compute(self) -> None:
+        # some operators and the data types have to be handled specially to align with pandas
+        # However, it is difficult to fail early if the arithmetic operator is not compatible
+        # with the data type, so we just let the server raise exception (e.g. a string minus a string).
+
+        if _op_is_between_two_timedeltas_or_timedelta_and_null(
```

Comment:
this looks confusing to me. I thought this case would be exclusive of `_op_is_between_timedelta_and_numeric`

Reply:
oh right, I just copy the original code here but seems this can be better.